probe-rs/src/debug/source_instructions.rs

```diff
@@ -23,6 +24,223 @@ pub struct VerifiedBreakpoint {
+                        &instruction_sequence,
+                        matching_file_index,
+                        line,
+                        column,
+                        debug_info,
+                        program_unit,
+                    )
+                    .or_else(|| {
+                        match_file_line_first_available_column(
+                            &instruction_sequence,
+                            matching_file_index,
+                            line,
+                            debug_info,
+                            program_unit,
+                        )
+                    }) {
+                        return Ok(verified_breakpoint);
+                    }
+                }
+            }
+        }
+        // If we get here, we have not found a valid breakpoint location.
+        Err(DebugError::Other(anyhow::anyhow!("No valid breakpoint information found for file: {}, line: {line:?}, column: {column:?}", path.to_path().display())))
+    }
+}
+
+/// Find the valid halt instruction location that is equal to, or greater than, the address.
+fn match_address(
+    instruction_sequence: &InstructionSequence<'_>,
+    address: u64,
+    debug_info: &DebugInfo,
+) -> Option<Result<VerifiedBreakpoint, DebugError>> {
+    if instruction_sequence.address_range.contains(&address) {
+        if let Some(valid_breakpoint) = instruction_sequence
+            .instructions
+            .iter()
+            .find(|instruction_location| {
+                instruction_location.instruction_type == InstructionType::HaltLocation
+                    && instruction_location.address >= address
+            })
+            .and_then(|instruction_location| {
+                SourceLocation::from_instruction_location(
+                    debug_info,
+                    instruction_sequence.program_unit,
+                    instruction_location,
+                )
+                .map(|source_location| VerifiedBreakpoint {
+                    address: instruction_location.address,
+                    source_location,
+                })
+            })
+        {
+            return Some(Ok(valid_breakpoint));
+        }
+    }
+    None
+}
+
+/// Find the valid halt instruction location that matches the file, line and column.
+fn match_file_line_column(
+    instruction_sequence: &InstructionSequence<'_>,
+    matching_file_index: Option<u64>,
+    line: u64,
+    column: Option<u64>,
+    debug_info: &DebugInfo,
+    program_unit: &UnitInfo,
+) -> Option<VerifiedBreakpoint> {
+    if let Some(instruction_location) =
+        instruction_sequence
+            .instructions
+            .iter()
+            .find(|instruction_location| {
+                instruction_location.instruction_type == InstructionType::HaltLocation
+                    && matching_file_index == Some(instruction_location.file_index)
+                    && NonZeroU64::new(line) == instruction_location.line
```

Comment:
```suggestion
                    && line == instruction_location.line.get()
```

let's not keep the possibility to panic :)

Reply:
Good catch ... thanks!

Reply:
@bugadani, this is not correct. the `instruction_location.line` is an `Option<NonZeroU64>`.

My `==` would agree on `None` values as well as `Some<u64>` values, which is exactly what I want. Why do you think this will panic?

Reply:
Oh nevermind then, brainfart on my part thinking that `::new()` panics, I forgot that it returns an Option.